libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2515,8 +2515,50 @@ void TypeChecker::requireLValue(Expression const& _expression)
 	_expression.annotation().lValueRequested = true;
 	_expression.accept(*this);
 
-	if (_expression.annotation().isConstant)
-		m_errorReporter.typeError(_expression.location(), "Cannot assign to a constant variable.");
-	else if (!_expression.annotation().isLValue)
-		m_errorReporter.typeError(_expression.location(), "Expression has to be an lvalue.");
+	if (_expression.annotation().isLValue)
+		return;
+
+	auto reasonOfFailure = [&]() -> const char* {
```

Comment:
If you prefer, I can really make this a member function, but I don't see it ever being used outside here and the lambda makes it much nicer already...

Reply:
To be honest, this makes the control flow clearer, but to really know what happens with the function, you have to scroll down. In my opinion, real modularity is only achieved if the lambda is taken out of the function. What about a free function in an anonymous namespace right above the member definition? Does the funciton need to know more than `_expression`?

Reply:
Ok, no problem :-). It calls ``TypeChecker::type``, but we could actually make that one static... I'll look into it.

Reply:
I changed the now second-to-last commit to inline the lambda-function into the error reporter call - personally I'd actually prefer that version and find it nicer than the free function, but I'm fine with that one as well and did it in the now latest commit (by also making ``TypeChecker::type`` static).